File: src/main/java/AoC2016_13.java
import com.github.pareronia.aoc.Grid.Cell;
import com.github.pareronia.aoc.StringOps;
import com.github.pareronia.aoc.Utils;
import com.github.pareronia.aoc.graph.Dijkstra;
import com.github.pareronia.aoc.solution.SolutionBase;

import java.util.List;
import java.util.function.Predicate;

@SuppressWarnings({"PMD.NoPackage", "PMD.ClassNamingConventions"})
public final class AoC2016_13 extends SolutionBase<Dijkstra.Result<Cell>, Integer, Integer> {

    private static final char ONE = '1';
    private static final Cell START = Cell.at(1, 1);

    private AoC2016_13(final boolean debug) {
        super(debug);
    }

    public static AoC2016_13 create() {
        return new AoC2016_13(false);
    }

    public static AoC2016_13 createDebug() {
        return new AoC2016_13(true);
    }

    @Override
    protected Dijkstra.Result<Cell> parseInput(final List<String> inputs) {
        final int input = Integer.parseInt(inputs.get(0));
        final Predicate<Cell> isOpenSpace =
                cell -> {
                    final int x = cell.getRow();
                    final int y = cell.getCol();
                    final int t = input + x * x + 3 * x + 2 * x * y + y + y * y;
                    final long ones =
                            Utils.asCharacterStream(Integer.toBinaryString(t))
                                    .filter(c -> c == ONE)
                                    .count();
                    return ones % 2 == 0;
                };
        return Dijkstra.execute(
                START,
                cell -> false,
                cell ->
                        cell.capitalNeighbours()
                                .filter(n -> n.getRow() >= 0 && n.getCol() >= 0)
                                .filter(isOpenSpace),
                (curr, next) -> 1);
    }

    private int getDistance(final Dijkstra.Result<Cell> input, final Cell end) {
        return (int) input.getDistance(end);
    }

    @Override
    public Integer solvePart1(final Dijkstra.Result<Cell> input) {
        return getDistance(input, Cell.at(31, 39));
    }

    @Override
    public Integer solvePart2(final Dijkstra.Result<Cell> input) {
        return (int) input.getDistances().values().stream().filter(v -> v <= 50).count();
    }

    @Override
    protected void samples() {
        final AoC2016_13 test = createDebug();
        final Dijkstra.Result<Cell> input = test.parseInput(StringOps.splitLines(TEST));
        assert test.getDistance(input, Cell.at(1, 1)) == 0;
        assert test.getDistance(input, Cell.at(7, 4)) == 11;
    }

    public static void main(final String[] args) throws Exception {
        create().run();
    }

    private static final String TEST = "10";
}

File: src/main/cpp/2016/13/AoC2016_13.cpp
#include <algorithm>
#include <bitset>

#include "../../aoc/graph/graph.hpp"
#include "../../aoc/grid/grid.hpp"

using namespace std;

const vector<int> DR = {-1, 0, 1, 0};
const vector<int> DC = {0, 1, 0, -1};
const Cell START = Cell::at(1, 1);

bool isOpenSpace(const uint input, const Cell& cell) {
    const int x = cell.row();
    const int y = cell.col();
    const int t = input + x * x + 3 * x + 2 * x * y + y + y * y;
    return bitset<32>(t).count() % 2 == 0;
}

unordered_set<Cell> findNeighbours(const uint input, const Cell& cell) {
    unordered_set<Cell> neighbours;
    for (uint i = 0; i < 4; i++) {
        const int rr = cell.row() + DR[i];
        const int cc = cell.col() + DC[i];
        const Cell n = Cell::at(rr, cc);
        if (rr >= 0 && cc >= 0 && isOpenSpace(input, n)) {
            neighbours.insert(n);
        }
    }
    return neighbours;
}

AStar::Result<Cell> runAStar(const vector<string>& input) {
    return AStar::execute<Cell>(
        START, [](const Cell& cell) { return false; },
        [&input](const Cell& cell) {
            return findNeighbours(stoi(input[0]), cell);
        },
        [](const Cell& cell) { return 1; });
}

int getDistance(const vector<string>& input, const Cell& end) {
    return runAStar(input).getDistances()[end];
}

int part1(const vector<string>& input) {
    return getDistance(input, Cell::at(31, 39));
}

int part2(const vector<string>& input) {
    const auto& dist = runAStar(input).getDistances();
    return count_if(dist.begin(), dist.end(),
                    [](const auto& item) { return item.second <= 50; });
}

const vector<string> TEST = {"10"};

void samples() {
    assert(getDistance(TEST, Cell::at(1, 1)) == 0);
    assert(getDistance(TEST, Cell::at(7, 4)) == 11);
}

MAIN(2016, 13)
